sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -577,26 +586,83 @@ public Instant timestamp(DoFn<InputT, OutputT> doFn) {
     switch (pTransform.getSpec().getUrn()) {
       case PTransformTranslation.SPLITTABLE_PROCESS_ELEMENTS_URN:
         this.convertSplitResultToWindowedSplitResult =
-            (splitResult, watermarkEstimatorState) ->
-                WindowedSplitResult.forRoots(
-                    WindowedValue.of(
-                        KV.of(
-                            currentElement.getValue(),
-                            KV.of(splitResult.getPrimary(), currentWatermarkEstimatorState)),
-                        currentElement.getTimestamp(),
-                        currentWindow,
-                        currentElement.getPane()),
-                    WindowedValue.of(
-                        KV.of(
-                            currentElement.getValue(),
-                            KV.of(splitResult.getResidual(), watermarkEstimatorState)),
-                        currentElement.getTimestamp(),
-                        currentWindow,
-                        currentElement.getPane()));
+            (splitResult, watermarkEstimatorState) -> {
+              List<BoundedWindow> primaryFullyProcessedWindows =
+                  ImmutableList.copyOf(
+                      Iterables.limit(
+                          currentElement.getWindows(), currentWindowIterator.previousIndex()));
+              // Advances the iterator consuming the remaining windows.
+              List<BoundedWindow> residualUnprocessedWindows =
+                  ImmutableList.copyOf(currentWindowIterator);
```

Comment:
`currentWindowIterator .next()`?

Reply:
ImmutableList.copyOf(iterator) drains all the remaining elements from iterator. It is effectively:
```
while (iterator.hasNext()) {
  list.add(iterator.next());
}
```

Reply:
Ah I see. It's different from what I know about an `Iterator`. Thanks!